src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/RazorHtmlPublishDiagnosticsInterceptor.cs

```diff
@@ -0,0 +1,195 @@
+            }
+
+            // Ensure we're working with the virtual document specified in the diagnostic params
+            if (!htmlDocumentSnapshot.Uri.Equals(diagnosticParams.Uri))
+            {
+                return CreateDefaultResponse(token);
+            }
+
+            // Note; this is an `interceptor` & not a handler, hence 
+            // it's possible another interceptor mutates this request 
+            // later in the toolchain. Such an interceptor would likely 
+            // expect a `__virtual.html` suffix instead of `.razor`.
+            diagnosticParams.Uri = razorDocumentUri;
+
+            // Return early if there aren't any diagnostics to process
+            if (diagnosticParams.Diagnostics?.Any() != true)
+            {
+                return CreateResponse(diagnosticParams);
+            }
+
+            var synchronized = await _documentSynchronizer.TrySynchronizeVirtualDocumentAsync(
+                razorDocumentSnapshot.Version,
+                htmlDocumentSnapshot,
+                cancellationToken).ConfigureAwait(false);
+            if (!synchronized)
+            {
+                // Could not synchronize, we'll clear out the diagnostics as we don't
+                // know whether or not they're still valid in the current state of the
+                // document.
+                return CreateResponse(diagnosticParams, clearDiagnostics: true);
+            }
+
+            var unmappedDiagnostics = diagnosticParams.Diagnostics;
+            var filteredDiagnostics = unmappedDiagnostics.Where(d => !CanDiagnosticBeFiltered(d));
+            if (!filteredDiagnostics.Any())
+            {
+                return CreateResponse(diagnosticParams, clearDiagnostics: true);
+            }
+
+            var rangesToMap = filteredDiagnostics.Select(r => r.Range).ToArray();
+            var mappingResult = await _documentMappingProvider.MapToDocumentRangesAsync(
+                RazorLanguageKind.Html,
+                razorDocumentUri,
+                rangesToMap,
+                LanguageServerMappingBehavior.Inclusive,
+                cancellationToken).ConfigureAwait(false);
+
+            if (mappingResult == null || mappingResult.HostDocumentVersion != razorDocumentSnapshot.Version)
+            {
+                return CreateResponse(diagnosticParams, clearDiagnostics: true);
+            }
+
+            var mappedDiagnostics = new List<Diagnostic>(filteredDiagnostics.Count());
+
+            for (var i = 0; i < filteredDiagnostics.Count(); i++)
+            {
+                var diagnostic = filteredDiagnostics.ElementAt(i);
+                var range = mappingResult.Ranges[i];
+
+                if (range.IsUndefined())
+                {
+                    // Couldn't remap the range correctly.
+                    // If this isn't an `Error` Severity Diagnostic we can discard it.
+                    if (diagnostic.Severity != DiagnosticSeverity.Error)
```

Comment:
So the criteria we're implementing is can't remap warnings means throw out but errors we keep?

Reply:
Exactly. This is the same logic we use for pull diagnostics. Reasoning being that error severity diagnostics likely break something catastrophic and we should always notify the user of the issue, regardless of whether or not we can actually map it successfully (which again for HTML should always be successful).

Reply:
This filtering logic will be moved to RZLS https://github.com/dotnet/aspnetcore/issues/27563 and consolidated for HTML/CSS & C# Pull Diagnostics 😄